### PR TITLE
Support intermediate polynomials

### PR DESCRIPTION
**Pull request by Schaeff**
- this reuses `FunctionDefinition` with a new variant to encode `col intermediate = e`
- it fails in witgen, to be fixed separately

### CODE REVIEW

File: ast/src/analyzed/mod.rs [hunk 1]
```diff
@@ -102,6 +108,7 @@ impl<T> Analyzed<T> {
             // for each kind.
             self.committed_polys_in_source_order(),
             self.constant_polys_in_source_order(),
+            self.intermediate_polys_in_source_order(),
```
Comment (chriseth) [inline ast/src/analyzed/mod.rs:111]: Please remove the TODO some lines up

File: pil_analyzer/src/pil_analyzer.rs [hunk 2]
```diff
@@ -651,6 +653,22 @@ namespace T(65536);
     col fixed p_reg_write_X_A = [0, 0, 0, 1, 0, 0, 0, 1, 0] + [0]*;
     col fixed p_reg_write_X_CNT = [1, 0, 0, 0, 0, 0, 0, 0, 0] + [0]*;
     { T.pc, T.reg_write_X_A, T.reg_write_X_CNT } in (1 - T.first_step) { T.line, T.p_reg_write_X_A, T.p_reg_write_X_CNT };
+"#;
+        let formatted = process_pil_file_contents::<GoldilocksField>(input).to_string();
+        if input != formatted {
+            for (i, f) in input.split('\n').zip(formatted.split('\n')) {
```
Comment (chriseth) [inline pil_analyzer/src/pil_analyzer.rs:659]: I think we could replace that by pretty_assert

Comment (Schaeff) [reply pil_analyzer/src/pil_analyzer.rs]: For this I changed to a more appropriate test based on the input and expected formatted output (which just adds the namespaces)

File: backend/src/pilcom_cli/json_exporter/mod.rs [hunk 3]
```diff
@@ -34,7 +34,7 @@ pub fn export<T: FieldElement>(analyzed: &Analyzed<T>) -> JsonValue {
             StatementIdentifier::Definition(name) => {
                 if let (poly, Some(value)) = &analyzed.definitions[name] {
                     if poly.poly_type == PolynomialType::Intermediate {
-                        if let FunctionValueDefinition::Mapping(value) = value {
```
Comment (leonardoalt) [inline backend/src/pilcom_cli/json_exporter/mod.rs:37]: so this was just buggy?

File: pil_analyzer/src/pil_analyzer.rs [hunk 4]
```diff
@@ -660,4 +662,20 @@ namespace T(65536);
         }
         assert_eq!(input, formatted);
     }
+
+    #[test]
+    fn intermediate() {
+        let input = r#"namespace N(65536);
+    col witness x;
+    col intermediate = x;
+    intermediate = intermediate;
```
Comment (leonardoalt) [inline pil_analyzer/src/pil_analyzer.rs:671]:
```suggestion
    intermediate = witness;
```
?

File: pil_analyzer/src/pil_analyzer.rs [hunk 5]
```diff
@@ -660,4 +662,20 @@ namespace T(65536);
...
+"#;
+        let expected = r#"namespace N(65536);
+    col witness x;
+    col intermediate = N.x;
+    N.intermediate = N.intermediate;
```
Comment (leonardoalt) [inline pil_analyzer/src/pil_analyzer.rs:676]: same here

File: pilopt/src/lib.rs [hunk 6]
```diff
@@ -411,6 +412,22 @@ mod test {
...
+        let optimized = optimize(process_pil_file_contents::<GoldilocksField>(input)).to_string();
+        assert_eq!(optimized, expectation);
+    }
+
+    #[test]
+    fn intermediate() {
+        let input = r#"namespace N(65536);
+        col witness x;
+        col intermediate = x;
+        intermediate = intermediate;
```
Comment (leonardoalt) [inline pilopt/src/lib.rs:425]:
```suggestion
        intermediate = witness;
```
?

File: pilopt/src/lib.rs [hunk 7]
```diff
@@ -411,6 +412,22 @@ mod test {
...
+    "#;
+        let expectation = r#"namespace N(65536);
+    col witness x;
+    col intermediate = N.x;
+    N.intermediate = N.intermediate;
```
Comment (leonardoalt) [inline pilopt/src/lib.rs:430]: same here?

File: test_data/asm/intermediate.asm [hunk 8]
```diff
@@ -0,0 +1,9 @@
+machine Intermediate(latch, operation_id) {
+    constraints {
+        col fixed latch = [1]*;
+        col fixed operation_id = [0]*;
+        col witness x;
+        col intermediate = x;
+        intermediate = intermediate;
```
Comment (leonardoalt) [inline test_data/asm/intermediate.asm:7]:
```suggestion
        intermediate = x;
```
?

**Review by leonardoalt — COMMENTED** [review]
looks good, just one question in tests

File: backend/src/pilcom_cli/json_exporter/mod.rs (hunk 3 above)
Comment (Schaeff) [reply backend/src/pilcom_cli/json_exporter/mod.rs:37]: I think so: it was treating `= x` the same as `(i) { x }` but this doesnt make sense from a type perspective: in the former `x` is a column (`int -> fe`) but in the latter (mapping used for fixed columns), x is a field element `fe`.

File: pil_analyzer/src/pil_analyzer.rs (hunk 4 above)
Comment (Schaeff) [reply pil_analyzer/src/pil_analyzer.rs:671]: 🤷

**Comment by leonardoalt** [comment]
> Actually, can you also add some tests with nested intermediates?

Ofc, just added.

**Review by leonardoalt — APPROVED** [review]
Looks good!

**Review by chriseth — APPROVED** [review]

**Pull request merged**